### PR TITLE
Move BuildDefinition from Rule to RuleDefinitionFactory

### LINKED ISSUE #315: Move Rule.GetDefinition to RuleDefinitionFactory
To conform to SRP. 
By moving this method, we will reduce number of dependencies in Rule class and will make RuleDefinitionFactory more meaningful

### PR DESCRIPTION
**Pull request by gritcsenko**
Since for every rule definition is built once, there is no need to cache its definition
Closes #315

### CODE REVIEW

**Comment by snikolayev** [comment]
@gritcsenko one thing that bothers me here is that the second call to `RuleDefinitionFactory.Create' for the same rule causes an exception due to duplicate declarations. I agree it probably doesn't make sense to cache definition, but I think anyone would expect the following to work, and create two equivalent definitions:
```c#
var rule = new TestRule();
var factory1 = new RuleDefinitionFactory();
var definition1 = factory1.Create(rule);
var factory2 = new RuleDefinitionFactory();
var definition1 = factory2.Create(rule);
```

I was thinking wrapping SymbolStack and RuleBuilder into a separate struct and using that as a mutable field on the Rule class. This way we would only create the SymbolStack and RuleBuilder when actually building the definition, and not when instantiating rules (so saving memory if a subset of rules is never compiled). And multiple factory invocations for the same rule would be possible.

```c#
internal readonly struct RuleBuildContext
{
    public static RuleBuildContext Empty = new RuleBuildContext();

    public SymbolStack SymbolStack { get; }
    public RuleBuilder Builder { get;}

    public RuleBuildContext(RuleBuilder builder)
    {
        SymbolStack = new SymbolStack();
        Builder = builder;
    }
}

public abstract class Rule
{
    private RuleBuildContext _context;
    
    //...

    internal void Define(RuleBuilder builder)
    {
        try
        {
            _context = new RuleBuildContext(builder);
            Define();
        }
        finally
        {
            _context = RuleBuildContext.Empty;
        }
    }
}
```

What do you think?

**Comment by gritcsenko** [comment]
@snikolayev I like this idea too. Let me try this one

**Comment by gritcsenko** [comment]
@snikolayev In order to make building definition safe I need to change visibility of `Define` method.

In case someone attempts to call public method `Rule.Define` it will fail w/o builder available.

**Comment by snikolayev** [comment]
> @snikolayev In order to make building definition safe I need to change visibility of `Define` method.
> 
> I case someone attempts to call public method `Rule.Define` it will fail w/o builder available.

@gritcsenko please don't change visibility of `Rule.Define` - that would be a huge breaking change. And while I agree it would be better if it was originally defined as protected, at this point it should remain as is. It's not reasonable to demand that everyone who uses NRules has to change every single rule they have written.

File: src/NRules/NRules.Fluent/Dsl/RuleBuildContext.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    public static RuleBuildContext CreateNew(RuleMetadata metadata)
+    {
+        var context = new RuleBuildContext(new RuleBuilder());
+        context.ApplyMetadata(metadata);
+        return context;
+    }
+
+    public static RuleBuildContext Empty = new();
+
+    public SymbolStack SymbolStack { get; } = new();
```
Comment (snikolayev) [inline src/NRules/NRules.Fluent/Dsl/RuleBuildContext.cs]: @gritcsenko wouldn't this instantiate an empty SymbolStack for Empty? I think the idea is that it's null for Empty

File: src/NRules/NRules.Fluent/RuleDefinitionFactory.cs [hunk 2]
```diff
@@ -36,11 +36,22 @@ public IRuleDefinition Create(Rule rule)
...
         }
     }
+
+    private IRuleDefinition BuildDefinition(Rule rule)
+    {
+        var metadata = new RuleMetadata(rule.GetType());
+
+        var context = RuleBuildContext.CreateNew(metadata);
+
+        rule.Define(context);
```
Comment (snikolayev) [inline src/NRules/NRules.Fluent/RuleDefinitionFactory.cs]: @gritcsenko I would instantiate the builder here, in the factory, and just pass to Rule.Define(builder). Then RuleBuilderContext is just an internal detail of the Rule class. The purpose of the RuleBuilderContext is mostly to wrap RuleBuilder and SymbolStack and it has no use outside of the Rule

File: src/NRules/NRules.Fluent/Dsl/RuleBuildContext.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    public RuleBuilder Builder { get; }
+
+    public bool IsEmpty => Builder is null;
+
+    private void ApplyMetadata(RuleMetadata metadata)
```
Comment (snikolayev) [inline src/NRules/NRules.Fluent/Dsl/RuleBuildContext.cs]: @gritcsenko I think it was better when it was inside the factory, no? The idea is that RuleBuilderContext is just a dumb wrapper, w/o any logic in it.

File: src/NRules/NRules.Fluent/Dsl/Rule.cs [hunk 4]
```diff
@@ -96,31 +95,23 @@ protected IRightHandSideExpression Then()
...
         {
-            _builder.Repeatability(metadata.Repeatability.Value);
+            _context = defaultContext;
         }
+    }
 
-        Define();
-
-        return _builder.Build();
+    private void InvalidateContext()
```
Comment (snikolayev) [inline src/NRules/NRules.Fluent/Dsl/Rule.cs]: @gritcsenko doesn't the name seem confusing? Invalidate means "make invalid". Probably makes more sense to call something like "ValidateContext".

Comment (gritcsenko) [reply src/NRules/NRules.Fluent/Dsl/Rule.cs]: My bad. Fixed

File: src/NRules/NRules.Fluent/Dsl/RuleBuildContext.cs (hunk 3 above)
Comment (gritcsenko) [reply src/NRules/NRules.Fluent/Dsl/RuleBuildContext.cs]: Moved, thanks for notice

File: src/NRules/NRules.Fluent/RuleDefinitionFactory.cs (hunk 2 above)
Comment (gritcsenko) [reply src/NRules/NRules.Fluent/RuleDefinitionFactory.cs]: Struct moved

File: src/NRules/NRules.Fluent/Dsl/RuleBuildContext.cs (hunk 1 above)
Comment (gritcsenko) [reply src/NRules/NRules.Fluent/Dsl/RuleBuildContext.cs]: Fixed

**Pull request merged**